File: backend/app/services/case_memory.py
from app.clients.mongo import get_db
from datetime import datetime


def save_draft(case_id: str, content: str) -> None:
	get_db()["drafts"].replace_one(
		{"case_id": case_id},
		{"case_id": case_id, "content": content, "saved_at": datetime.utcnow()},
		upsert=True,
	)


def get_latest_draft(case_id: str) -> str | None:
	doc = get_db()["drafts"].find_one({"case_id": case_id}, {"_id": 0})
	return doc["content"] if doc else None


def save_research(case_id: str, sources: list) -> None:
	get_db()["research"].replace_one(
		{"case_id": case_id},
		{"case_id": case_id, "sources": sources, "cached_at": datetime.utcnow()},
		upsert=True,
	)


def get_cached_research(case_id: str) -> list | None:
	doc = get_db()["research"].find_one({"case_id": case_id}, {"_id": 0})
	return doc["sources"] if doc else None

File: frontend/lib/services/caseMemory.ts
import { getDb } from '@/lib/mongo';

export const saveDraft = async (caseId: string, content: string): Promise<void> =>
{
	const db = await getDb();
	await db.collection('drafts').replaceOne(
		{ case_id: caseId },
		{ case_id: caseId, content, saved_at: new Date() },
		{ upsert: true }
	);
};

export const getLatestDraft = async (caseId: string): Promise<string | null> =>
{
	const db = await getDb();
	const doc = await db.collection('drafts').findOne({ case_id: caseId }, { projection: { _id: 0 } });
	return doc ? (doc.content as string) : null;
};

export const saveResearch = async (caseId: string, sources: any[]): Promise<void> =>
{
	const db = await getDb();
	await db.collection('research').replaceOne(
		{ case_id: caseId },
		{ case_id: caseId, sources, cached_at: new Date() },
		{ upsert: true }
	);
};

export const getCachedResearch = async (caseId: string): Promise<any[] | null> =>
{
	const db = await getDb();
	const doc = await db.collection('research').findOne({ case_id: caseId }, { projection: { _id: 0 } });
	return doc ? (doc.sources as any[]) : null;
};
